### ISSUE TITLE
srv/ui layout with jetty webserver : jetty deps are packaged into the war

### ISSUE THREAD

**Issue by fcamblor** [restx/restx#220]
Currently, when we're generating srv/ui layout, we have `restx-server-jetty` dependency added with `compile` scope

Problem with this scope is that this dependency (and all its transitive deps like `servlet3.jar`) will be packaged into the war's `WEB-INF/lib` directory.
I consider it as a bug.

The problem is not easy to solve.
We could use `provided` scope here (thus, libs wouldn't be packaged into the war, and jetty classes would be provided during compilation time). However, in tools like Intellij, when we launch the `AppServer`, we run into `ClassNotFoundException: restx.server.JettyWebServer` errors as Intellij don't provide `provided` libs at runtime.
Some issues may be found related to this topic in youtrack : [here](https://youtrack.jetbrains.com/issue/IDEA-66392#tab=Linked%20Issues), [there](https://youtrack.jetbrains.com/issue/IDEABKL-99), [there](https://youtrack.jetbrains.com/issue/IDEA-107048) and [this discussion](https://devnet.jetbrains.com/message/5466609)

Another way to solve the issue would be to use a maven profile to add this dependency only on a `dev` maven profile. But I didn't got it to work on my Intellij version (15 EAP currently), it seems like maven activated profiles are not taken into consideration when it comes to building the classpath for the java app runner.

Another way would be to add [proper jar file exclusions](http://maven.apache.org/plugins/maven-war-plugin/examples/including-excluding-files-from-war.html) in maven-war-plugin configuration.
It may be hard to maintain over time, but we're not changing jetty dependency everydays either...

Last thing working for me : putting an `<optional>true</optional>` on the `restx-server-jetty` dependency in pom.xml.
On my side, it seems to work (no transitive deps packaged into my war, and when I run the `AppServer`, it works)

**Comment by fcamblor** [comment]
Ok I confirm the `<optional>true</optional>` workaround seems to work from dev to deploy (in a tomcat6 env)

For history, after I put the `<optional>true</option>` I had to manually add the `io.restx:restx-servlet` artefact (it was a transitive dep pulled by `restx-server-jetty` and which was required at runtime, with its `janino` transitive deps as well)


The only thing I'm not completely satisfied of is that, at compile time, we're grabbing lots of servlet-apis from `restx-server-jetty` transitive deps :
- javax.servlet:javax.servlet-api (in version 3.0), which is considered a different artefact than `javax.servlet:servlet-api` (no `javax` in artefactId) grabbed from `restx-servlet` (version 2.5) with `provided` scope
- org.eclipse.jetty.orbit:javax.servlet (in version 3.0 as well)

It means that, in the IDE, developper may reference these classes which won't be available once deployed into a non-jetty webserver.

I don't know if we can workaround this, I tried to put exclusions on the `restx-server-jetty` artefact : 
```
        <dependency>
            <groupId>io.restx</groupId>
            <artifactId>restx-server-jetty</artifactId>
            <version>${restx.version}</version>
            <optional>true</optional>
            <exclusions>
                <exclusion>
                    <groupId>javax.servlet</groupId>
                    <artifactId>javax.servlet-api</artifactId>
                </exclusion>
                <exclusion>
                    <groupId>org.eclipse.jetty.orbit</groupId>
                    <artifactId>javax.servlet</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
```
however, by doing this, we cannot reference any `servlet-api` classes in our cases unless we add something like :
```
        <dependency>
            <groupId>javax.servlet</groupId>
            <artifactId>servlet-api</artifactId>
            <version>2.5</version>
            <scope>provided</scope>
        </dependency>
```
And by doing this, the code compiles but we get `NoClassDefFoundError: javax/servlet/http/HttpServletResponse` when running the `AppServer` from Intellij (as exposed in my previous post, `provided` deps are not added at runtime by Intellij)

So I think we're in a dead end here, the only solution would be to provide a `restx-server-jetty7` (or 6) component for servlet 2.5 api.

I will make a PR with the `<optional>true</optional>` + `io.restx:restx-servlet` artefact declaration.
We'll see later if this is really needed to make a `jetty7` server to be sure we don't have servlet 3.0-specific stuff in developper code.

**Comment by xhanin** [comment]
I really miss Ivy configurations in such cases... I think using the optional approach is ok. And if someone need to use the servlet api directly indeed they will need to add it as direct dependency, but I don't see why it breaks when launching `AppServer`, doesn't the `optional` dependency on `jetty` bring the servlet3 api in classpath?

Or maybe we should simply consider levelling up the dependency on servlet API. I think the most widely used app server in prod for restx is tomcat, and tomcat supports servlet 3 since version 7, version 6 is tied to 2.5 but the team has announced an end of life for 31 december 2016: http://tomcat.apache.org/tomcat-60-eol.html

**Comment by fcamblor** [comment]
but I don't see why it breaks when launching AppServer, doesn't the optional dependency on jetty bring the servlet3 api in classpath?

Yes, optional dependency *without* dependency exclusions is working fine (if we don't consider we have servlet-api 3.0 classes available at compile time in the IDE)
Problems are coming when we try to exclude this extra dependency (I'm wondering, in that case, if it shouldn't be an interesting feature, in Intellij, to allow to bring `provided` libs in runtime classpath when explicitely requested)

To my POV, 31th of december 2016 is a long period of time, I'm not sure dropping servlet 2.5 support would be a good news for everyone.

**Comment by xhanin** [comment]
Ok, I thought you were saying that adding the direct dependency was breaking the AppServer run (because you mentioned the issue with `provided`), but it's the exclusions that breaks it, so we can't have the exclusion. So the only clean solution to that issue I can see would be to have support for an embedded servlet 2.5 server, like tomcat 6 or jetty 7. I'm not sure tomcat6 can be easily used as embedded, but jetty 7 should be. That way we can stick to servlet 2.5 requirement in `restx-servlet` module, and have people use it safely by depending on the `restx-server-jetty7` module. We'd need to rename `restx-server-jetty` to `restx-server-jetty8` to be cleaner, I think it would be better anyway, we don't want to tie a particular restx version to a particular server version, we can support several of them. But that's more work, maybe we can just add the `optional` for the moment.

WDYT?

**Comment by fcamblor** [comment]
@xhanin I merged PR #221 and created #227 to not forget about creating a new jetty webserver artefact for servlet 2.5
